leakpro/attacks/utils/boosting.py

```diff
@@ -0,0 +1,212 @@
+        The privacy score definition can be found in "Membership Inference Attacks From First Principles" by Carlini et al.
+        """
+
+        logits_function = ModelRescaledLogits()
+        logits = np.swapaxes(logits_function(self.shadow_models, self.handler, self.audit_data_indices, self.batch_size), 0, 1)
+        target_logits = np.swapaxes(logits_function([self.target_model], self.handler, self.audit_data_indices, self.batch_size),\
+                                    0, 1).squeeze()
+
+        privacy_score = []
+
+        # From "Membership Inference Attacks From First Principles", fixed variance is used when number of shadow models < 64
+        if len(self.shadow_models) < 64:
+            in_std, out_std = np.std(logits[self.in_indices_mask].flatten()), np.std(logits[~self.in_indices_mask].flatten())
+
+        for (logit, target_logit, mask) in tqdm(zip(logits, target_logits, self.in_indices_mask),
+                                                total=len(logits),
+                                                desc="Calculating privacy score",
+                                                leave=False
+                                                ):
+            in_mean, out_mean = np.mean(logit[mask]), np.mean(logit[~mask])
+
+            if len(self.shadow_models) >= 64:
+                in_std, out_std = np.std(logit[mask]), np.std(logit[~mask])
+
+            if self.online:
+                privacy_score.append(np.abs(in_mean-out_mean)/(in_std+out_std+1e-30))
+            else:
+                privacy_score.append(np.abs(target_logit-out_mean)/(2*out_std+1e-30))
+
+        self.privacy_score = np.asarray(privacy_score)
+
+    def softmax_logits(self:Self, logits: np.ndarray) -> np.ndarray:
+        """Rescale logits to (0, 1).
+
+        Args:
+        ----
+            logits ( len(dataset) x ... x nb_classes ): Logits to be rescaled.
+
+        """
+        logits = torch.from_numpy(logits)
+        logits = logits - torch.max(logits, dim=-1, keepdim=True).values
+        logits = torch.exp(logits)
+        logits = logits/torch.sum(logits, dim=-1, keepdim=True)
+        return logits.numpy()
+
+    def adjust_memorization_mask(self:Self) -> list:
+        """Adjust thesholds to achieve the desired amount or percentile of most vulnerable datapoints."""
+
+        # Set starting thresholds, from the paper ("why train more...")
+        mem_thrshld = 0.8
+        priv_thrshld = 2.0
+
+        # Use the thresholds from the paper
+        if self.memorization_threshold == 0.0:
+            return self.memorization_score > mem_thrshld, self.privacy_score > priv_thrshld
+
+        # Adjust initial thresholds if they are set too high
+        while (np.count_nonzero((self.memorization_score > mem_thrshld)\
+                | (self.privacy_score > priv_thrshld)) < self.num_memorization_audit_points):
+            mem_thrshld = mem_thrshld/2
```

Comment:
should these be lowered in sequence rather than simultaneously?

Reply:
I think we have to reason around and argue what metric is more important if we should do this. The same reason for why I set the increment to 1.001 for both, setting a higher increment for privacy score showed better performance for the tests I did, but I feel like I dont wanna argue for that disparity